HW1/P6/P6.py

```diff
@@ -0,0 +1,219 @@
+    cap_per = only_capitals_period.match(word)
+    
+    # if it's a number...
+    numbs = only_numbers.match(word)
+        
+    result = True
+    if (caps) or (cap_per) or (numbs):
+        result = False
+        
+    return result
+
+def count_occurrences(words):
+    """ Takes an iterable, words, and returns a list
+    of tuples of each word that appears in it and the number of times it appears.
+    To be used with map after groupByKey.
+    """
+    
+    results = []
+    dist_words = set(words)
+    orig_words = list(words)
+    
+    for word in dist_words:
+        c = orig_words.count(word)
+        results.append((word, c))
+        
+    return results
+
+def generate_phrase(markov, num_words):
+    """Takes a markov chain, assumed to be an RDD of the form described in the problem
+    and returns a randomly generated phrase with num_words words. The phrase is generated by
+    choosing a pair of words randomly from the RDD, weighting the next word by the relative
+    frequencies of occurrences of other words after that pair in Shakespeare, and then
+    continues to do so until we have num_words.
+    """
+    # Our counter
+    curr_num_words = 0
+    
+    # Start us off with a blank phrase
+    phrase = []
+
+    # First flag is for with replacement... doesn't matter because we just do one at a time
+    # Note that takeSample returns a list, so we just really want one element from it
+    next_three = markov.takeSample(True, 1)[0]
+    
+    # Get the first two words
+    word_1 = next_three[0][0]
+    word_2 = next_three[0][1]
+    
+    # Now get the list of possible words
+    poss_words = next_three[1]
+    
+    # And choose accordingly...
+    word_3 = weighted_choice(poss_words)
+    
+    # Add our words to the phrase...
+    phrase.append(word_1)
+    phrase.append(word_2)
+    phrase.append(word_3)
+    
+    # Make sure we stop at some point
+    curr_num_words += 3
+
+    # Now our next pair is the previous two words
+    next_pair = (word_2, word_3)
+
+    # And just keep iterating until we are done 
+    while (curr_num_words < num_words):
+        # Get the list of possible words
+        # Do this weird thing because of the Spark error
+        next_poss_words = markov.map(lambda x: x).lookup(next_pair)[0]
+        
+        # Get the actual next word
+        next_word = weighted_choice(next_poss_words)
+
+        # Add it to the phrase
+        phrase.append(next_word)
+
+        # And shift the pair
+        next_pair = (next_pair[1], next_word)
+
+        curr_num_words += 1
+
+                
+    # Now we get it to be a string
+    phrase = ' '.join(phrase)
+    
+    # And we are done!
+    return phrase
+
+def weighted_choice(choices):
+    """ Inspired by http://stackoverflow.com/questions/3679694/a-weighted-version-of-random-choice
+    Given a list of tuples, [(x1, c1), (x2, c2), ...]
+    Returns some xn from the list chosen randomly but weighted by the counts cn.
+    """
+    
```

Comment:
Another option here is to use `zip` to get the counts and then divide the counts by the length of the list, which gives you the probabilities you need.